test/common/upstream/health_check_fuzz.cc

```diff
@@ -112,34 +111,402 @@ void HealthCheckFuzz::triggerTimeoutTimer(bool last_action) {
+  grpc::health::v1::HealthCheckResponse response;
+  response.set_status(status);
+  const auto data = Grpc::Common::serializeToGrpcFrame(response);
+  auto ret = std::vector<uint8_t>(data->length(), 0);
+  data->copyOut(0, data->length(), &ret[0]);
+  return ret;
+}
+
+Buffer::OwnedImpl GrpcHealthCheckFuzz::makeBufferToRespondWith(
+    test::common::upstream::GrpcRespondBytes grpc_respond_bytes) {
+  switch (grpc_respond_bytes.grpc_respond_bytes_selector_case()) {
+  case test::common::upstream::GrpcRespondBytes::kStatus: {
+    // Structured Response
+    std::vector<uint8_t> data;
+    switch (grpc_respond_bytes.status()) {
+    case test::common::upstream::ServingStatus::UNKNOWN: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::UNKNOWN);
+      break;
+    }
+    case test::common::upstream::ServingStatus::SERVING: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::SERVING);
+      break;
+    }
+    case test::common::upstream::ServingStatus::NOT_SERVING: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::NOT_SERVING);
+      break;
+    }
+    case test::common::upstream::ServingStatus::SERVICE_UNKNOWN: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::SERVICE_UNKNOWN);
+      break;
+    }
+    default: { // shouldn't hit
+      NOT_REACHED_GCOVR_EXCL_LINE;
+      break;
+    }
+    }
+    Buffer::OwnedImpl response;
+    response.add(data.data(), data.size());
+    return response;
+  }
+  case test::common::upstream::GrpcRespondBytes::kData: {
+    // Arbitrarily Generated Bytes
+    Buffer::OwnedImpl response;
+    response.add(grpc_respond_bytes.data());
+
+    ENVOY_LOG_MISC(trace,
+                   "Responded with {}. Length (in bytes) = {}. These bytes were arbitrarily "
+                   "generated and have no structure.",
+                   grpc_respond_bytes.data(), grpc_respond_bytes.data().length());
+
+    return response;
+  }
+  default: { // shouldn't hit
+    NOT_REACHED_GCOVR_EXCL_LINE;
+    break;
+  }
+  }
+}
+
+// Logic from respondResponseSpec() in unit tests
+void GrpcHealthCheckFuzz::respond(test::common::upstream::GrpcRespond grpc_respond) {
+  if (!test_sessions_[0]->timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+  // These booleans help figure out when to end the stream
+  bool has_data = false;
+  bool has_trailers = false;
+
+  if (grpc_respond.has_grpc_respond_bytes()) {
+    switch (grpc_respond.grpc_respond_bytes().grpc_respond_bytes_selector_case()) {
+    case test::common::upstream::GrpcRespondBytes::kStatus: {
+      has_data = true;
+      break;
+    }
+    case test::common::upstream::GrpcRespondBytes::kData: {
+      if (grpc_respond.grpc_respond_bytes().data().length() != 0) {
+        has_data = true;
+      }
+      break;
+    }
+    default: // If the fuzzer generates an empty grpc_respond_bytes message
```

Comment:
Added validation, switched this line to the macro of not hit.

Reply:
This logic still needs to be split, because an empty respond_bytes logically represents an empty response (from what I remember from TCP Health checking).

Reply:
"This logic still needs to be split, because an empty respond_bytes"

wouldn't an empty respond_bytes still be call to decodeData with an empty buffer? as long as that case can be captured, i'm happy with a solution!

Reply:
Oh yeah sorry you're totally right! I was thinking of bytes in the health check config and conflated it with bytes in response! This can simplify my check down to one if! Thanks!

Reply:
Switched to:
if (grpc_respond.has_grpc_respond_bytes()) {
    has_data = true;
}